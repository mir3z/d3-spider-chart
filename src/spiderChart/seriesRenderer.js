import * as d3 from "d3";
import noop from "lodash.noop";

/**
 * Renders each data series.
 * Data series can be one of type: 'area', 'line', 'dot'.
 * Clicking series displays values near each data point
 * @param parent SVG parent group
 * @param innerRadius inner radius at which series starts
 * @param outerRadius outer radius at which series ends
 * @param color color scale to colorize each series
 * @returns {function}
 */
export default function seriesRenderer(parent, { innerRadius, outerRadius, color }) {
    const group = parent
        .append("g")
        .attr("class", CLASS_NAME.SERIES_CONTAINER);

    const radialAreaGenerator = d3.areaRadial()
        .angle(d => d.angle)
        .curve(d3.curveLinearClosed)
        .innerRadius(d => d.innerRadius)
        .outerRadius(d => d.outerRadius);

    const radialLineGenerator = d3.lineRadial()
        .angle(d => d.angle)
        .curve(d3.curveLinearClosed)
        .radius(d => d.outerRadius);

    const calculateDataPoints = (radiusScale) => (d) => d.values
        .map((v, i) => ({
            angle: 2 * Math.PI / d.values.length * i,
            innerRadius: innerRadius,
            outerRadius: radiusScale(v.value),
            value: v.value,
            type: d.type,
            key: d.key
        }));

    const setSeriesPolygon = (radiusScale) => el => el
        .attr("d", d => {
            const points = calculateDataPoints(radiusScale)(d);

            return d.type === "area"
                ? radialAreaGenerator(points)
                : radialLineGenerator(points);
        })
        .attr("fill", (d, i) => d.type === "area" ? color(i) : "none")
        .attr("stroke", (d, i) => d.type === "line" ? color(i) : "none");

    const handleSeriesClick = (data, update) => d => {
        const nextData = data
            .map(series => ({
                ...series,
                focus: series.key === d.key ? !series.focus : false,
                values: series.key === d.key
                    ? series.values.map(v => ({ ...v, focus: !series.focus, highlight: false, dim: false }))
                    : series.values.map(v => ({ ...v, focus: false, highlight: false, dim: false }))
            }));

        update(nextData);
    };

    return function render(data = [], update = noop) {
        const maxValue = d3.max(data, s => d3.max(s.values, v => v.value));
        const radiusScale = d3.scaleLinear()
            .domain([0, maxValue])
            .range([innerRadius, outerRadius * SCALE_FACTOR]);

        const seriesGroup = group
            .selectAll(`.${ CLASS_NAME.SERIES_GROUP }`)
            .data(data)
            .classed(CLASS_NAME.SERIES_GROUP_HIDDEN, d => d.hidden);

        seriesGroup.exit().remove();

        const enterSeries = seriesGroup.enter()
            .append("g")
            .attr("class", CLASS_NAME.SERIES_GROUP)
            .attr("data-series", d => d.key);

        enterSeries
            .append("path")
            .attr("class", CLASS_NAME.SERIES_POLYGON)
            .call(setSeriesPolygon(radiusScale))
            .on("click", handleSeriesClick(data, update));

        seriesGroup.select(`.${ CLASS_NAME.SERIES_POLYGON }`)
            .call(setSeriesPolygon(radiusScale))
            .on("click", handleSeriesClick(data, update));

        const seriesDots = enterSeries
            .attr("fill", (d, i) => color(i))
            .selectAll(`.${ CLASS_NAME.SERIES_POLYGON_DOT }`)
            .data(d => {
                if (d.type === "area") {
                    return false;
                }

                return calculateDataPoints(radiusScale)(d);
            });

        seriesDots.enter()
            .append("circle")
            .attr("class", CLASS_NAME.SERIES_POLYGON_DOT)
            .attr("cx", d => d.outerRadius * Math.sin(d.angle))
            .attr("cy", d => -d.outerRadius * Math.cos(d.angle))
            .attr("r", 4)
            .on("click", handleSeriesClick(data, update));

        seriesGroup.selectAll(`.${ CLASS_NAME.SERIES_POLYGON_DOT }`)
            .data(d => {
                if (d.type === "area") {
                    return false;
                }

                return calculateDataPoints(radiusScale)(d);
            })
            .attr("cx", d => d.outerRadius * Math.sin(d.angle))
            .attr("cy", d => -d.outerRadius * Math.cos(d.angle))
            .attr("r", 4)
            .on("click", handleSeriesClick(data, update));


        setLabelsPlacement(group, getLabelNodes(data, radiusScale, color));

        return group;
    };
}

const getLabelNodes = (data, radiusScale, color, nodeLabelOffset = 12) => data
    .flatMap((d, j) =>
        d.hidden
            ? []
            : d.values.map((v, i) => ({
                ...v,
                angle: 2 * Math.PI / d.values.length * i,
                color: color(j),
                outerRadius: radiusScale(v.value) + nodeLabelOffset
            }))
    )
    .filter(v => v.focus);

const setLabelsPlacement = (group, nodes, steps = 500) => {
    const simulation = d3.forceSimulation(nodes)
        .force("charge", d3.forceManyBody().strength(-10))
        .force("x", d3.forceX().x(d => d.outerRadius * Math.sin(d.angle)))
        .force("y", d3.forceY().y(d => -d.outerRadius * Math.cos(d.angle)))
        .force("collision", d3.forceCollide().radius(18))
        .stop();

    for (let i = 0; i < steps; i++) {
        simulation.tick();
    }

    const labels = group
        .selectAll(`.${ CLASS_NAME.SERIES_VALUE }`)
        .data(nodes);

    labels.enter()
        .append("text")
        .merge(labels)
        .attr("class", CLASS_NAME.SERIES_VALUE)
        .attr("data-axis", d => d.axis)
        .attr("fill", d => d.color)
        .attr("x", d => d.x)
        .attr("y", d => d.y)
        .text(d => d.value);

    labels.exit().remove();
};

const SCALE_FACTOR = 0.85;
const CLASS_NAME = {
    SERIES_CONTAINER: "series",
    SERIES_GROUP: "series-group",
    SERIES_GROUP_HIDDEN: "series-group--hidden",
    SERIES_POLYGON: "series-polygon",
    SERIES_POLYGON_DOT: "series-polygon-dot",
    SERIES_VALUE: "series-value"
};
