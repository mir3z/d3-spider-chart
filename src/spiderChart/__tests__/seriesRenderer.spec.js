import "core-js/fn/array/flat-map";
import * as d3 from "d3";
import seriesRenderer from "../seriesRenderer";
import { cleanup, setup } from "./utils";

describe("series renderer", () => {
    const color = d3.scaleLinear()
        .domain([0, 1, 2, 3, 4])
        .range([
            "#53c6ff",
            "#0021ba",
            "#ff8823",
            "#187a2e"
        ])
        .interpolate(d3.interpolateHcl);

    const defaultOptions = {
        color,
        innerRadius: 10,
        outerRadius: 100,
    };

    const data = [
        {
            key: "Series 1",
            type: "area",
            focus: true,
            values: [
                { axis: "Statement A", group: "Group 1", value: 20, focus: true },
                { axis: "Statement B", group: "Group 2", value: 22, focus: true },
                { axis: "Statement C", group: "Group 3", value: 24, focus: true },
            ]
        },
        {
            key: "Series 2",
            type: "line",
            values: [
                { axis: "Statement A", group: "Group 1", value: 10 },
                { axis: "Statement B", group: "Group 2", value: 15 },
                { axis: "Statement C", group: "Group 3", value: 20 },
            ]
        },
        {
            key: "Series 3",
            type: "dot",
            values: [
                { axis: "Statement A", group: "Group 1", value: 11 },
                { axis: "Statement B", group: "Group 2", value: 22 },
                { axis: "Statement C", group: "Group 3", value: 33 },
            ]
        },
    ];

    afterEach(cleanup);

    it("renders its own group", () => {
        const svg = setup();
        const renderSeries = seriesRenderer(svg, defaultOptions);

        const group = renderSeries(data);
        const ownGroup = svg.node().querySelector("g.series");

        expect(ownGroup).not.toBeNull();
        expect(group.node()).toEqual(ownGroup);
    });

    it("correctly renders all series", () => {
        const svg = setup();
        const renderSeries = seriesRenderer(svg, defaultOptions);

        const group = renderSeries(data);

        expect(group.node()).toMatchSnapshot();
    });

    it("calls update when series is clicked", () => {
        const svg = setup();
        const update = jest.fn();
        const renderSeries = seriesRenderer(svg, defaultOptions);

        const group = renderSeries(data, update);
        group.node().querySelector("[data-series='Series 3'] path")
            .dispatchEvent(new Event("click"));

        expect(update).toHaveBeenCalledWith([
            {
                key: "Series 1",
                type: "area",
                focus: false,
                values: [
                    { axis: "Statement A", group: "Group 1", value: 20, focus: false, highlight: false, dim: false },
                    { axis: "Statement B", group: "Group 2", value: 22, focus: false, highlight: false, dim: false },
                    { axis: "Statement C", group: "Group 3", value: 24, focus: false, highlight: false, dim: false },
                ]
            },
            {
                key: "Series 2",
                type: "line",
                focus: false,
                values: [
                    { axis: "Statement A", group: "Group 1", value: 10, focus: false, highlight: false, dim: false },
                    { axis: "Statement B", group: "Group 2", value: 15, focus: false, highlight: false, dim: false },
                    { axis: "Statement C", group: "Group 3", value: 20, focus: false, highlight: false, dim: false },
                ]
            },
            {
                key: "Series 3",
                type: "dot",
                focus: true,
                values: [
                    { axis: "Statement A", group: "Group 1", value: 11, focus: true, highlight: false, dim: false },
                    { axis: "Statement B", group: "Group 2", value: 22, focus: true, highlight: false, dim: false },
                    { axis: "Statement C", group: "Group 3", value: 33, focus: true, highlight: false, dim: false },
                ]
            },
        ]);
    });

    it("re-renders when data changes", () => {
        const svg = setup();
        const renderSeries = seriesRenderer(svg, defaultOptions);

        renderSeries(data);
        const newData = [
            {
                key: "Series 1",
                type: "area",
                values: [
                    { axis: "Statement A", group: "Group 1", value: 10 },
                    { axis: "Statement B", group: "Group 2", value: 11 },
                    { axis: "Statement C", group: "Group 3", value: 12 },
                ]
            }
        ];
        const group = renderSeries(newData);

        expect(group.node()).toMatchSnapshot();
    });
});
