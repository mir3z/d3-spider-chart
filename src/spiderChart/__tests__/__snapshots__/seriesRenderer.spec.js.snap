// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`series renderer correctly renders all series 1`] = `
<g
  class="series"
>
  <g
    class="series-group"
    data-series="Series 1"
    fill="rgb(83, 198, 255)"
  >
    <path
      class="series-polygon"
      d="M0,-55.45454545454545L51.96152422706632,29.999999999999986L-55.89800333517739,32.2727272727273ZM-8.660254037844386,5.000000000000004L8.660254037844387,4.999999999999998L0,-10Z"
      fill="rgb(83, 198, 255)"
      stroke="none"
    />
  </g>
  <g
    class="series-group"
    data-series="Series 2"
    fill="rgb(0, 33, 186)"
  >
    <path
      class="series-polygon"
      d="M0,-32.72727272727273L38.183847348677524,22.045454545454533L-48.02504511895523,27.72727272727275Z"
      fill="none"
      stroke="rgb(0, 33, 186)"
    />
    <circle
      class="series-polygon-dot"
      cx="0"
      cy="-32.72727272727273"
      r="4"
    />
    <circle
      class="series-polygon-dot"
      cx="38.183847348677524"
      cy="22.045454545454533"
      r="4"
    />
    <circle
      class="series-polygon-dot"
      cx="-48.02504511895523"
      cy="27.72727272727275"
      r="4"
    />
  </g>
  <g
    class="series-group"
    data-series="Series 3"
    fill="rgb(255, 136, 35)"
  >
    <path
      class="series-polygon"
      d="M0,-35L51.96152422706632,29.999999999999986L-73.61215932167727,42.500000000000036Z"
      fill="none"
      stroke="none"
    />
    <circle
      class="series-polygon-dot"
      cx="0"
      cy="-35"
      r="4"
    />
    <circle
      class="series-polygon-dot"
      cx="51.96152422706632"
      cy="29.999999999999986"
      r="4"
    />
    <circle
      class="series-polygon-dot"
      cx="-73.61215932167727"
      cy="42.500000000000036"
      r="4"
    />
  </g>
  <text
    class="series-value"
    data-axis="Statement A"
    fill="rgb(83, 198, 255)"
    x="0.017607981637723515"
    y="-68.80149254456771"
  >
    20
  </text>
  <text
    class="series-value"
    data-axis="Statement B"
    fill="rgb(83, 198, 255)"
    x="63.513155160449905"
    y="36.69009403828773"
  >
    22
  </text>
  <text
    class="series-value"
    data-axis="Statement C"
    fill="rgb(83, 198, 255)"
    x="-67.46790356226639"
    y="38.92532809641399"
  >
    24
  </text>
</g>
`;

exports[`series renderer re-renders when data changes 1`] = `
<g
  class="series"
>
  <g
    class="series-group"
    data-series="Series 1"
    fill="rgb(83, 198, 255)"
  >
    <path
      class="series-polygon"
      d="M0,-72.5L68.19950054802455,39.374999999999986L-73.61215932167727,42.500000000000036ZM-8.660254037844386,5.000000000000004L8.660254037844387,4.999999999999998L0,-10Z"
      fill="rgb(83, 198, 255)"
      stroke="none"
    />
  </g>
</g>
`;
